src/wp-includes/theme.php

```diff
@@ -188,23 +191,31 @@ function get_stylesheet() {
  *
  * @since 1.5.0
  *
+ * @global string $wp_stylesheet_path Current theme stylesheet directory path.
+ *
  * @return string Path to active theme's stylesheet directory.
  */
 function get_stylesheet_directory() {
-	$stylesheet     = get_stylesheet();
-	$theme_root     = get_theme_root( $stylesheet );
-	$stylesheet_dir = "$theme_root/$stylesheet";
+	global $wp_stylesheet_path;
 
-	/**
-	 * Filters the stylesheet directory path for the active theme.
-	 *
-	 * @since 1.5.0
-	 *
-	 * @param string $stylesheet_dir Absolute path to the active theme.
-	 * @param string $stylesheet     Directory name of the active theme.
-	 * @param string $theme_root     Absolute path to themes directory.
-	 */
-	return apply_filters( 'stylesheet_directory', $stylesheet_dir, $stylesheet, $theme_root );
+	if ( null === $wp_stylesheet_path ) {
+		$stylesheet     = get_stylesheet();
+		$theme_root     = get_theme_root( $stylesheet );
+		$stylesheet_dir = "$theme_root/$stylesheet";
+
+		/**
+		 * Filters the stylesheet directory path for the active theme.
+		 *
+		 * @since 1.5.0
+		 *
+		 * @param string $stylesheet_dir Absolute path to the active theme.
+		 * @param string $stylesheet     Directory name of the active theme.
+		 * @param string $theme_root     Absolute path to themes directory.
+		 */
+		$wp_stylesheet_path = apply_filters( 'stylesheet_directory', $stylesheet_dir, $stylesheet, $theme_root );
```

Comment:
Using a global like this mean that the filters will only fire once. This maybe a breaking change.

Reply:
Correct, the filter will fire once per page load, except when the active theme has changed since the last time (because of the global being reset).

I agree with you that there is a small risk here, but using the filter to return a different theme directory for a single theme being active is very unlikely. I think we should post a draft note to explain this change (same for the `template_directory` filter) in case someone is using it in weird ways. It's still possible to force the filter to re-run by manually setting the global back to `null`. Obviously not recommended in general, but if there's a very specific use-case, it would be possible.

Reply:
What if the filter is added after the first time that `get_stylesheet_directory` is run. Like if it was run again init or plugins_loaded. I am pretty sure that `get_stylesheet_directory` is called very early in the bootstrap.